### PR TITLE
Use Eloquent models

### PR DESCRIPTION
**Pull request by NorbiPeti**
- Moved the Knight model to App\Model
- Fixed Doctrine MySQL bit type handling for artisan commands, so we can generate model IDE helper classes
- Created models for the data we query from the DB and started using them instead of raw SQL
- Added 'active' global scope to automatically limit queries to what the user can see
- Added method and constants to be able to have a generic way of handling permissions (also a bit more verbose but more understandable)
- Added support for update/create dates (Laravel)
- Set up Xdebug automatically to work in Docker

### CODE REVIEW

File: app/Http/Controllers/ProfileController.php [hunk 1]
```diff
@@ -331,7 +333,7 @@ public function update(Request $request, $rname)
 
                 // Add skills, reactivate deleted ones if they exist
                 foreach ($added as $skill) {
-                    $prev_deleted = $knight->skills()->wherePivot('delflg', true)->find($skill);
+                    $prev_deleted = $knight->skills(true)->find($skill);
```
Comment (LeoVerto) [inline app/Http/Controllers/ProfileController.php:336]: I'd prefer explicity using the named argument in situations like this one to make the code more readable.

File: app/Http/Controllers/ProfileController.php [hunk 2]
```diff
@@ -360,9 +362,9 @@ public function update(Request $request, $rname)
                     $divModel->pivot->save();
                 }
 
-                // Add skills, reactivate deleted ones if they exist
+                // Add divisions, reactivate deleted ones if they exist
                 foreach ($added as $div) {
-                    $prev_deleted = $knight->divisions()->wherePivot('delflg', true)->find($div);
+                    $prev_deleted = $knight->divisions(true)->find($div);
```
Comment (LeoVerto) [inline app/Http/Controllers/ProfileController.php:367]: Same as above.

**Review by LeoVerto — COMMENTED** [review]
Looking pretty good so far! This modernisation was much-needed, the benefits of using raw SQL queries never really materialized.

I like the addition of global traits for active/deleted. That's certainly much easier than explicitly handling them everywhere in the code.

I'm not sure what the right move is regarding moving the interfaces out of User and directly into Knight. This could mean additional maintenance work because of upstream changes in the future.

File: app/Http/Controllers/ProfileController.php (hunk 1 above)
Comment (NorbiPeti) [reply app/Http/Controllers/ProfileController.php]: That's fair, added the name of the parameter.

File: app/Http/Controllers/ProfileController.php (hunk 2 above)
Comment (NorbiPeti) [reply app/Http/Controllers/ProfileController.php]: Same here.

**Comment by NorbiPeti** [comment]
> I'm not sure what the right move is regarding moving the interfaces out of User and directly into Knight. This could mean additional maintenance work because of upstream changes in the future.

I agree, but they don't seem to be any large or complicated implementations, so I want to go in this direction. I'm removing the interfaces related to email login (CanResetPassword, MustVerifyEmail). The Authenticatable interface is talked about in the docs, I couldn't find anything for the Authorizable interface (though atm we aren't using any Laravel authorization AFAIK anyway). Personally, I don't think these implementations are going anywhere anytime soon but I could be wrong. In any case, we can deal with it when we next update Laravel lol.

Another option would be to convert my model superclass into a trait and an interface, though I don't really like that concept here. I feel like things like the primary key column name shouldn't be in a trait (though the permission/delflg stuff could be).

**Review by LeoVerto — APPROVED** [review]
Okay, let's go ahead and get this merged. 

We'll see whether building our own User class causes problems during the upcoming migration to Laravel 9 and rolling it back wouldn't be too hard anyways.

Thanks for putting all this work into getting the codebase up to speed!

**Pull request merged**